### PR TITLE
add act-as powers

### PR DESCRIPTION
**Pull request by deads2k**
This adds the ability for a user to Act-As another user if he has the ability to verb="impersonate", resource="users", resourceName=`<value of Act-As Header>`.

@erictune @kubernetes/kube-iam @liggitt @smarterclayton 

I'll need to update the authorizer to handle resource names through the interface, but I don't plan to update the abac policy file.

### CODE REVIEW

**Comment by deads2k** [comment]
I think I've found the section with auth tests in the integration suite.  I'll see about adding a test there.

File: pkg/apiserver/handlers.go [hunk 1]
```diff
@@ -413,6 +415,65 @@ func (r *requestAttributeGetter) GetAttribs(req *http.Request) authorizer.Attrib
...
+			newUser := &user.DefaultInfo{
+				Name: requestedSubject,
+			}
+			requestContextMapper.Update(req, api.WithUser(ctx, newUser))
+		}
+
+		newCtx, _ := requestContextMapper.Get(req)
+		oldUser, _ := api.UserFrom(ctx)
+		newUser, _ := api.UserFrom(newCtx)
+		glog.V(3).Infof("%v is acting as %v on %v", oldUser, newUser, req.URL)
```
Comment (erictune) [inline pkg/apiserver/handlers.go]: Is this request going to go into the HTTP log as the original or the impersonated user?  I think the original user is more useful from a debugging and audit logging perspective, so we should try to preserve that.  If it is possible to decorate the logs lines with the impersonated user name, that would be nice, but not as important as logging the original user.

Comment (liggitt) [inline pkg/apiserver/handlers.go]: isn't it logging both the original and new user?

Comment (liggitt) [inline pkg/apiserver/handlers.go]: from this point on, anything that pulls the user out of the context will get the impersonated user

Comment (erictune) [inline pkg/apiserver/handlers.go]: I was talking about the logs that the go http server emits to stdout and go into `/var/log/kube-apiserver.log` on a typical kubernetes cluster.  That doesn't use `glog`.

Comment (erictune) [inline pkg/apiserver/handlers.go]: I thought that was logging the username [here](https://github.com/kubernetes/kubernetes/blob/v1.2.0/pkg/httplog/log.go) but it is just logging the User-Agent header.  So, never mind.

**Comment by erictune** [comment]
looks good to me, but let lavalamp review too.

File: test/integration/auth_test.go [hunk 2]
```diff
@@ -725,6 +725,89 @@ func TestUnknownUserIsUnauthorized(t *testing.T) {
...
+	if a.GetUserName() == "bob" && a.GetVerb() == "act-as" {
+		return nil
+	}
+	return errors.New("I can't allow that.  Go ask alice.")
+}
+
+func TestActAsIsForbidden(t *testing.T) {
+	framework.DeleteAllEtcdKeys()
+
+	// This file has alice and bob in it.
```
Comment (lavalamp) [inline test/integration/auth_test.go]: What file?

File: test/integration/auth_test.go [hunk 3]
```diff
@@ -725,6 +725,89 @@ func TestUnknownUserIsUnauthorized(t *testing.T) {
...
+
+	for _, r := range getTestRequests() {
+		token := BobToken
+		bodyBytes := bytes.NewReader([]byte(r.body))
+		req, err := http.NewRequest(r.verb, s.URL+r.URL, bodyBytes)
+		if err != nil {
+			t.Fatalf("unexpected error: %v", err)
+		}
+		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
+		req.Header.Set("Act-As", "alice")
```
Comment (lavalamp) [inline test/integration/auth_test.go]: Does the system just trust Bob's determination that the end user was really alice?

File: pkg/client/unversioned/clientcmd/api/v1/types.go [hunk 4]
```diff
@@ -82,6 +82,8 @@ type AuthInfo struct {
 	ClientKeyData []byte `json:"client-key-data,omitempty"`
 	// Token is the bearer token for authentication to the kubernetes cluster.
 	Token string `json:"token,omitempty"`
+	// ActAs is the username to act-as.
+	ActAs string `json:"act-as,omitempty"`
```
Comment (lavalamp) [inline pkg/client/unversioned/clientcmd/api/v1/types.go]: Seems unlikely we'd actually want this set in a config file?

File: pkg/client/transport/round_trippers.go [hunk 5]
```diff
@@ -130,6 +133,36 @@ func (rt *basicAuthRoundTripper) CancelRequest(req *http.Request) {
 
 func (rt *basicAuthRoundTripper) WrappedRoundTripper() http.RoundTripper { return rt.rt }
 
+type actAsRoundTripper struct {
+	actAs string
+	rt    http.RoundTripper
+}
+
+// NewActAsRoundTripper will apply a BASIC auth authorization header to a
```
Comment (lavalamp) [inline pkg/client/transport/round_trippers.go]: Comment seems wrong.

File: pkg/client/transport/round_trippers.go [hunk 6]
```diff
@@ -130,6 +133,36 @@ func (rt *basicAuthRoundTripper) CancelRequest(req *http.Request) {
 
 func (rt *basicAuthRoundTripper) WrappedRoundTripper() http.RoundTripper { return rt.rt }
 
+type actAsRoundTripper struct {
+	actAs string
+	rt    http.RoundTripper
```
Comment (lavalamp) [inline pkg/client/transport/round_trippers.go]: please call this "delegate"

File: pkg/apiserver/handlers.go [hunk 7]
```diff
@@ -413,6 +415,65 @@ func (r *requestAttributeGetter) GetAttribs(req *http.Request) authorizer.Attrib
...
+			requestContextMapper.Update(req, api.WithUser(ctx, serviceaccount.UserInfo(namespace, name, "")))
+
+		default:
+			newUser := &user.DefaultInfo{
+				Name: requestedSubject,
+			}
+			requestContextMapper.Update(req, api.WithUser(ctx, newUser))
+		}
+
+		newCtx, _ := requestContextMapper.Get(req)
```
Comment (lavalamp) [inline pkg/apiserver/handlers.go]: Don't ignore error?

File: pkg/apiserver/handlers.go (hunk 1 above)
Comment (lavalamp) [inline pkg/apiserver/handlers.go]: Please do `httplog.LogOf(req, w).Addf("%v is acting as %v", oldUser, newUser)` to actually associate this message with the request in addition to logging directly here.

Comment (liggitt) [inline pkg/apiserver/handlers.go]: That's nifty… didn't know about that

Comment (smarterclayton) [inline pkg/apiserver/handlers.go]: We unwrap that thing so often I'm surprised we still keep it around.

File: test/integration/auth_test.go (hunk 3 above)
Comment (deads2k) [inline test/integration/auth_test.go]: > Does the system just trust Bob's determination that the end user was really alice?

Bob's power to do that is controlled based on whether the API server authorizer says that Bob has the power to "act-as", the resource "users", with name "Alice".  For this test, its set up with this mock here: https://github.com/kubernetes/kubernetes/pull/23549/files/85844c419933c4a507225d52bfa542f8922685e1#diff-8754bfba92df8f58b6e09a4200975a90R735 .

File: pkg/client/unversioned/clientcmd/api/v1/types.go (hunk 4 above)
Comment (deads2k) [inline pkg/client/unversioned/clientcmd/api/v1/types.go]: > Seems unlikely we'd actually want this set in a config file?

I wasn't sure.  One significant use case is an `su` style interaction where you have an admin who doesn't normally have the power to perform certain actions, but can escalate himself when he needs to.  For one offs, he'd use `--as=`, but why shouldn't he be allowed to create a profile with it if he wants?  I wasn't planning to make it easy, just possible (and it is today using `kubectl config set`.

If you'd prefer, I can pull it out of the serialized format, but I'd like to keep it on the internal format to allow cleaner object merging.

File: pkg/apiserver/handlers.go (hunk 7 above)
Comment (deads2k) [inline pkg/apiserver/handlers.go]: > Don't ignore error?
>  Add a line note

It's an `exists` bool for a context that we just set a few lines up

File: pkg/apiserver/handlers.go [hunk 8]
```diff
@@ -413,6 +415,65 @@ func (r *requestAttributeGetter) GetAttribs(req *http.Request) authorizer.Attrib
...
+				forbidden(w, req)
+				return
+			}
+			requestContextMapper.Update(req, api.WithUser(ctx, serviceaccount.UserInfo(namespace, name, "")))
+
+		default:
+			newUser := &user.DefaultInfo{
+				Name: requestedSubject,
+			}
+			requestContextMapper.Update(req, api.WithUser(ctx, newUser))
```
Comment (deads2k) [inline pkg/apiserver/handlers.go]: And I ignored this one because we just did a `Get` (and checked `exists`) a few lines up from here.

**Comment by deads2k** [comment]
Two things:
 1. We never settled on the header name: `Act-As`, `X-Authentication-Acting-As`, `Impersonate-User`?  I like `Act-As` best, but that might just be because that's how the original proposal was titled.

 2. The authorization check for service accounts (https://github.com/kubernetes/kubernetes/pull/23549/files#diff-6073dcd8ae44859aa2bee1af7531dfe0R437).  It's a synthetic check, so we can make it check whatever we want.  I chose the simple implementation for this pull ("users"), but I'd like to expand it allow a check for `act-as` against the exact `serviceaccounts` resource when an SA user is specified.  I think that more accurately expresses what they're trying to do.  If this is non-contentious I'll do it here.  If we need to talk about it an issue, we can do that so we don't block the utility in this pull.

**Comment by erictune** [comment]
Apparently "X-" is deprecated for headers.  "Act-As" is so terse that I think it is unlikely to ever become a standard for this sort of thing, and doesn't document itself if someone is trying to understand our protocol by looking at the headers.  "Impersonation" is sort of a term-of-art, and it is suggestive that the header relates to auth. So, I vote for David's third option, `Impersonate-User`.

**Comment by liggitt** [comment]
Agree. I would change the verb check to "impersonate" as well

**Comment by erictune** [comment]
Fine with `impersonate` verb as well.

For your item 2, David, are you suggesting something like this for a service account:

```go
	actingAsAttributes := &authorizer.AttributesRecord{
			User:     requestor,
			Verb:     "act-as",
			APIGroup: api.GroupName,
			Resource: "serviceAccounts",
			ResourceName:    serviceAccountName,
			ResourceRequest: true,
		}
```
And possibly code to verify existence of the service account object in question in the namespace?

File: pkg/apiserver/handlers.go [hunk 9]
```diff
@@ -413,6 +415,65 @@ func (r *requestAttributeGetter) GetAttribs(req *http.Request) authorizer.Attrib
...
+			forbidden(w, req)
+			return
+		}
+
+		actingAsAttributes := &authorizer.AttributesRecord{
+			User:     requestor,
+			Verb:     "act-as",
+			APIGroup: api.GroupName,
+			Resource: "users",
+			// ResourceName:    requestedSubject,
```
Comment (erictune) [inline pkg/apiserver/handlers.go]: If I approve the "GetExtra" PR the would you put requestor user.Info.GetExtras["full-username"] or something like that?

Comment (deads2k) [inline pkg/apiserver/handlers.go]: > If I approve the "GetExtra" PR the would you put requestor user.Info.GetExtras["full-username"] or something like that?

I think I'd like to namespace the key, but I'm ok with that.  Would you prefer that to filling in something more structured on the context itself?

**Comment by deads2k** [comment]
> For your item 2, David, are you suggesting something like this for a service account:

Yes.



> And possibly code to verify existence of the service account object in question in the namespace?

I don't know that I'd force existence.  I could imagine it being useful to impersonate someone who can't actually log in themselves.

**Comment by deads2k** [comment]
Updated to `Impersonate-User` and updated code to match, but left the flag as `--as`, which is shorter and I think it expresses intent well.

**Comment by liggitt** [comment]
> left the flag as --as, which is shorter and I think it expresses intent well.

I like it

**Comment by erictune** [comment]
:+1:

**Comment by lavalamp** [comment]
LGTM

**Comment by erictune** [comment]
@cjcullen FYI

**Comment by erictune** [comment]
@roberthbailey FYI

**Pull request merged**